plugin/restore/mysql/mysql.go

```diff
@@ -511,9 +520,113 @@ func binlogFilesAreContinuous(files []BinlogFile) bool {
-func (r *Restore) FetchArchivedBinlogFiles(ctx context.Context) error {
+func (r *Restore) reconcileLocalBinlogFiles(ctx context.Context, binlogFilesLocalSorted, binlogFilesOnServerSorted []BinlogFile) error {
+	// Re-download inconsistent local binlog files.
+	// This is a best-effort task, because the corresponding binlog files may not exist on server.
+	latestBinlogFileOnServer := binlogFilesOnServerSorted[len(binlogFilesOnServerSorted)-1]
+	log.Debug("Re-downloading inconsistent local binlog files")
+	binlogFilesOnServerMap := make(map[string]BinlogFile)
+	for _, file := range binlogFilesOnServerSorted {
+		binlogFilesOnServerMap[file.Name] = file
+	}
+	for _, file := range binlogFilesLocalSorted {
+		fileOnServer, existOnServer := binlogFilesOnServerMap[file.Name]
+		if existOnServer && file.Size != fileOnServer.Size {
+			path := filepath.Join(r.binlogDir, file.Name)
+			log.Debug("Deleting inconsistent local binlog file",
+				zap.String("path", path),
+				zap.Int64("sizeLocal", file.Size),
+				zap.Int64("sizeOnServer", fileOnServer.Size))
+			if err := os.Remove(path); err != nil {
+				log.Error("Failed to remove inconsistent local binlog file", zap.String("path", path), zap.Error(err))
+				return fmt.Errorf("failed to remove inconsistent local binlog file[%s], error[%w]", path, err)
+			}
+			if err := r.downloadBinlogFile(ctx, fileOnServer, file.Name == latestBinlogFileOnServer.Name); err != nil {
+				log.Error("Failed to re-download inconsistent local binlog file", zap.String("path", path), zap.Error(err))
+				return fmt.Errorf("failed to re-download inconsistent local binlog file[%s], error[%w]", path, err)
+			}
+		}
+	}
+	return nil
+}
+
+func (r *Restore) canSkipDownloadBinlogFilesOnServer(ctx context.Context, binlogFilesLocalSorted, binlogFilesOnServerSorted []BinlogFile, targetTs int64) (skip bool, err error) {
+	// Check whether we can only use local binlog files to restore to targetTs:
+	// 1. The local binlog files have no gaps.
+	// 2. Some file's first binlog event ts is already larger than targetTs.
+	// 3. The first local binlog file <= first server binlog file
+	// If all true, we can skip downloading binlog files at all.
+	log.Debug("Checking whether we can just replay existing local binlog files to restore")
+	if binlogFilesLocalSorted[0].Seq <= binlogFilesOnServerSorted[0].Seq {
+		isContinuous := binlogFilesAreContinuous(binlogFilesLocalSorted)
+		if isContinuous {
+			for _, file := range binlogFilesLocalSorted {
+				eventTs, err := r.parseFirstLocalBinlogEventTimestamp(ctx, file.Name)
+				path := filepath.Join(r.binlogDir, file.Name)
+				if err != nil {
+					log.Error("Failed to parse the first binlog event timestamp for local binlog file",
+						zap.String("path", path),
+						zap.Error(err))
+					return false, fmt.Errorf("failed to parse the first binlog event timestamp for local binlog file[%s], error[%w]", path, err)
+				}
+				if eventTs > targetTs {
```

Comment:
What's the reason to get the start time of each binary log? We've already converted the input targetTs to the binary log position which can be used to compare.

We should always convert things to binary position (filename, position) and do the comparison as a single way. Don't compare time and position.

Reply:
We have not converted targetTs to binlog coordinate (filename, position).
We are converting binlog coordinates in the backup payload to eventTs.
In short, we are converting things to ts and comparing ts.

If we want to change to comparing binlog coordinates, there should be a re-design and refactor.

Reply:
And converting targetTs to binlog coordinate also requires parsing at least starting eventTs of each binlog file (even repeatedly if we do not maintain ts of the first/last event for archived binlog files).
It's like:
1. parse binlog files into ts ranges
2. find the belonging ts range of the targetTs, which gives the corresponding binlog file